### PR TITLE
New module: Text en-/decoding

### PR DESCRIPTION
**Pull request by ennru**
This module offers flows for character set encoding, decoding and transcoding.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -162,6 +163,8 @@ lazy val sqs = alpakkaProject("sqs",
 
 lazy val sse = alpakkaProject("sse", "sse", Dependencies.Sse)
 
+lazy val text = alpakkaProject("text", "text")
```
Comment (2m) [inline build.sbt]: Move down a bit.

File: text/src/main/scala/akka/stream/alpakka/text/impl/CharsetDecodingFlow.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import java.nio.charset.Charset
+
+import akka.stream.stage.{GraphStage, GraphStageLogic}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.util.ByteString
+
+/**
+ * Decodes a stream of bytes into a stream of characters, using a supplied {@link Charset}.
+ */
+private[text] class CharsetDecodingFlow(incoming: Charset) extends GraphStage[FlowShape[ByteString, String]] {
```
Comment (2m) [inline text/src/main/scala/akka/stream/alpakka/text/impl/CharsetDecodingFlow.scala]: InternalApi

File: text/src/main/scala/akka/stream/alpakka/text/impl/CharsetLogic.scala [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+import java.nio.charset.{Charset, CharsetDecoder, CharsetEncoder}
+
+import akka.stream.{FlowShape, Inlet, Outlet}
+import akka.stream.stage.{GraphStageLogic, InHandler, OutHandler}
+import akka.util.ByteString
+
+/**
+ * ByteBuffer to CharBuffer decoding logic.
+ */
+private[impl] trait Decoding {
```
Comment (2m) [inline text/src/main/scala/akka/stream/alpakka/text/impl/CharsetLogic.scala]: InternalApi here and below in the file.

File: build.sbt (hunk 1 above)
Comment (ennru) [reply build.sbt]: What do you mean?

**Comment by Voltir** [comment]
@ennru - I guess its been awhile since I last checked on the state of alpakka, but better late then never? This library does solve a good chunk of the problem around text encodings and is quite useful in its own right, so thank you for that. 

I still think its problematic that alpakka-csv uses ByteString and Bytes though. - I think there is room in the ecosystem for "CharStrings" and that the various text parsing transforms should really rely on that as a primitive. For example, even today alpakka-csv can't handle certain delimiters, even if the stream is already UTF-8 encoded. At the end of the day, UTF-8 is a variable length encoding where a single character can be encoded with more than one byte - and alpakka-csv and friends are operating on too low level of a primitive to deal with that.

**Comment by ennru** [comment]
I agree, the byte-based parsing is not optimal and doing it char-based is much more correct.
On the other hand, this solves most use cases well without introducing a new wrapper type like `CharString`. Even going from a `ByteString` to a `CharString` will require translating the whole thing properly into chars and you might get some left-over bytes.

**Comment by Voltir** [comment]
@ennru One thing we did to help in that regard was write a utility flow that attempts to canonicalize ByteStrings (assumed to be text in an unknown encoding) into UTF-8. After testing a number of libraries, we settled on using this com.ibm.icu.text.CharsetDetector as the basis for doing the detection. Would a flow like that be something alpakka is interested in?

Also, if we did want to go further and explore something like a "CharString" - where do you think something like that would belong? It seems like its not quite at the level of akka, with its akka.util.ByteString, but its also not quite at the alpakka csv library level either.

**Comment by ennru** [comment]
The `com.ibm.icu.text.CharsetDetector`-base flow would be a great addition to this Alpakka Text module.

There is no obvious home for the `CharString`, you're right. Making a jar just for that seems not great either, but that way both Alpakka Text and CSV could depend on it.

**Pull request merged**